File: backend/voidswitch/api/admin/proxies.py
"""Admin: outbound proxy pool management."""

from __future__ import annotations

import datetime as dt

from fastapi import APIRouter, Depends, HTTPException, Request, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from voidswitch.constants import ProxyStatus
from voidswitch.core.audit import record_audit
from voidswitch.core.auth import actor_display_name, require_staff
from voidswitch.core.database import get_session
from voidswitch.models.db import Provider, Proxy, User
from voidswitch.models.schemas import ProxyCreate, ProxyOut, ProxyUpdate
from voidswitch.services.network import Route, probe_route

router = APIRouter(prefix="/api/admin/proxies", tags=["admin:proxies"])


@router.get("", response_model=list[ProxyOut])
async def list_proxies(
    session: AsyncSession = Depends(get_session),
    _: User = Depends(require_staff),
) -> list[Proxy]:
    rows = (await session.execute(select(Proxy).order_by(Proxy.id))).scalars().all()
    return list(rows)


@router.post("", response_model=list[ProxyOut], status_code=status.HTTP_201_CREATED)
async def add_proxies(
    body: ProxyCreate,
    request: Request,
    session: AsyncSession = Depends(get_session),
    user: User = Depends(require_staff),
) -> list[Proxy]:
    existing = {u for (u,) in (await session.execute(select(Proxy.url))).all()}
    created: list[Proxy] = []
    seen: set[str] = set()
    for raw in body.urls:
        url = raw.strip()
        if not url or url in existing or url in seen:
            continue
        seen.add(url)
        proxy = Proxy(
            url=url,
            local_address=body.local_address,
            weight=body.weight,
            note=body.note,
            status=ProxyStatus.ACTIVE.value,
        )
        session.add(proxy)
        created.append(proxy)
    if not created:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "No new proxies to add.")
    await session.flush()
    await record_audit(
        session,
        action="proxy.add",
        actor_sub=user.sub,
        actor_name=user.name,
        target_type="proxy",
        detail={"added": len(created)},
        ip=request.client.host if request.client else None,
    )
    return created


@router.patch("/{proxy_id}", response_model=ProxyOut)
async def update_proxy(
    proxy_id: int,
    body: ProxyUpdate,
    request: Request,
    session: AsyncSession = Depends(get_session),
    user: User = Depends(require_staff),
) -> Proxy:
    proxy = await session.get(Proxy, proxy_id)
    if proxy is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Proxy not found.")
    changes = body.model_dump(exclude_unset=True)
    if changes.get("enabled") is True or changes.get("status") == ProxyStatus.ACTIVE.value:
        proxy.failed_count = 0
        proxy.disabled_reason = None
        proxy.status = ProxyStatus.ACTIVE.value
    for field, value in changes.items():
        if field == "enabled":
            continue
        setattr(proxy, field, value)
    if "enabled" in changes:
        proxy.enabled = changes["enabled"]
    await session.flush()
    await record_audit(
        session,
        action="proxy.update",
        actor_sub=user.sub,
        actor_name=actor_display_name(user),
        target_type="proxy",
        target_id=proxy_id,
        detail={"changes": list(changes)},
        ip=request.client.host if request.client else None,
    )
    return proxy


@router.delete("/{proxy_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_proxy(
    proxy_id: int,
    request: Request,
    session: AsyncSession = Depends(get_session),
    user: User = Depends(require_staff),
) -> None:
    proxy = await session.get(Proxy, proxy_id)
    if proxy is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Proxy not found.")
    # Scrub this id from any provider's selected proxy_ids so no dangling
    # reference is left behind. (proxy_ids is JSON — there's no FK to cascade.)
    providers = (await session.execute(select(Provider))).scalars().all()
    for p in providers:
        if p.proxy_ids and proxy_id in p.proxy_ids:
            p.proxy_ids = [pid for pid in p.proxy_ids if pid != proxy_id]
    await record_audit(
        session,
        action="proxy.delete",
        actor_sub=user.sub,
        actor_name=actor_display_name(user),
        target_type="proxy",
        target_id=proxy_id,
        detail={"url": proxy.url},
        ip=request.client.host if request.client else None,
    )
    await session.delete(proxy)


@router.post("/{proxy_id}/probe", response_model=ProxyOut)
async def probe_proxy(
    proxy_id: int,
    request: Request,
    session: AsyncSession = Depends(get_session),
    user: User = Depends(require_staff),
) -> Proxy:
    proxy = await session.get(Proxy, proxy_id)
    if proxy is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Proxy not found.")
    route = Route(proxy_url=proxy.url or None, local_address=proxy.local_address)
    ok, latency, _status, error = await probe_route(route, "https://api.openai.com/v1/models")
    proxy.last_checked_at = dt.datetime.now(dt.UTC)
    proxy.latency_ms = latency
    if ok:
        proxy.status = ProxyStatus.ACTIVE.value
        proxy.failed_count = 0
        proxy.disabled_reason = None
    else:
        proxy.disabled_reason = error
    await session.flush()
    await record_audit(
        session,
        action="proxy.probe",
        actor_sub=user.sub,
        actor_name=actor_display_name(user),
        target_type="proxy",
        target_id=proxy_id,
        detail={"ok": ok, "latency_ms": latency},
        ip=request.client.host if request.client else None,
    )
    return proxy

File: backend-go/internal/handlers/admin/proxies.go
package admin

import (
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"

	"github.com/siiway/voidswitch/internal/core"
	"github.com/siiway/voidswitch/internal/database"
	"github.com/siiway/voidswitch/internal/models"
	"github.com/siiway/voidswitch/internal/services"
)

func listProxies(c *gin.Context) {
	db := database.GetDatabase().DB
	user, ok := getCurrentUserGin(c)
	if !ok || !requireStaffGin(c, user) {
		return
	}

	var proxies []database.Proxy
	if err := db.Find(&proxies).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to query proxies."})
		return
	}

	result := make([]models.ProxyOut, 0, len(proxies))
	for _, p := range proxies {
		result = append(result, proxyToOut(p))
	}

	c.JSON(http.StatusOK, result)
}

func batchAddProxies(c *gin.Context) {
	db := database.GetDatabase().DB
	user, ok := getCurrentUserGin(c)
	if !ok || !requireStaffGin(c, user) {
		return
	}

	var body models.ProxyCreate
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body."})
		return
	}

	if len(body.URLs) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "No URLs provided."})
		return
	}

	name := core.ActorDisplayName(user)
	var created []models.ProxyOut

	for _, url := range body.URLs {
		if url == "" {
			continue
		}
		record := database.Proxy{
			URL:          url,
			LocalAddress: body.LocalAddress,
			Weight:       body.Weight,
			Note:         body.Note,
		}
		if err := db.Create(&record).Error; err != nil {
			continue
		}
		created = append(created, proxyToOut(record))
	}

	core.RecordAudit(db, "proxy.create", &user.Sub, &name, nil, nil,
		map[string]any{"count": len(created)}, nil, nil, nil, "admin")

	c.JSON(http.StatusOK, created)
}

func updateProxy(c *gin.Context) {
	db := database.GetDatabase().DB
	user, ok := getCurrentUserGin(c)
	if !ok || !requireStaffGin(c, user) {
		return
	}

	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid proxy ID."})
		return
	}

	var existing database.Proxy
	if err := db.First(&existing, id).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Proxy not found."})
		return
	}

	var body models.ProxyUpdate
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body."})
		return
	}

	changes := make(map[string]any)

	if body.URL != nil {
		changes["url"] = map[string]any{"old": existing.URL, "new": *body.URL}
		existing.URL = *body.URL
	}
	if body.LocalAddress != nil {
		changes["local_address"] = map[string]any{"old": existing.LocalAddress, "new": *body.LocalAddress}
		existing.LocalAddress = body.LocalAddress
	}
	if body.Enabled != nil {
		changes["enabled"] = map[string]any{"old": existing.Enabled, "new": *body.Enabled}
		existing.Enabled = *body.Enabled
	}
	if body.Status != nil {
		changes["status"] = map[string]any{"old": existing.Status, "new": *body.Status}
		existing.Status = *body.Status
	}
	if body.Weight != nil {
		changes["weight"] = map[string]any{"old": existing.Weight, "new": *body.Weight}
		existing.Weight = *body.Weight
	}
	if body.Note != nil {
		changes["note"] = map[string]any{"old": existing.Note, "new": *body.Note}
		existing.Note = body.Note
	}

	if err := db.Save(&existing).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update proxy."})
		return
	}

	cname := core.ActorDisplayName(user)
	targetID := strconv.Itoa(existing.ID)
	targetType := "proxy"
	core.RecordAudit(db, "proxy.update", &user.Sub, &cname, &targetType, &targetID, changes, nil, nil, nil, "admin")

	c.JSON(http.StatusOK, proxyToOut(existing))
}

func deleteProxy(c *gin.Context) {
	db := database.GetDatabase().DB
	user, ok := getCurrentUserGin(c)
	if !ok || !requireStaffGin(c, user) {
		return
	}

	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid proxy ID."})
		return
	}

	var existing database.Proxy
	if err := db.First(&existing, id).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Proxy not found."})
		return
	}

	if err := db.Delete(&existing).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete proxy."})
		return
	}

	cname := core.ActorDisplayName(user)
	targetID := strconv.Itoa(id)
	targetType := "proxy"
	core.RecordAudit(db, "proxy.delete", &user.Sub, &cname, &targetType, &targetID,
		map[string]any{"url": existing.URL}, nil, nil, nil, "admin")

	c.Status(http.StatusNoContent)
}

func probeProxy(c *gin.Context) {
	db := database.GetDatabase().DB
	user, ok := getCurrentUserGin(c)
	if !ok || !requireStaffGin(c, user) {
		return
	}

	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid proxy ID."})
		return
	}

	var existing database.Proxy
	if err := db.First(&existing, id).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Proxy not found."})
		return
	}

	probeURL := services.GetCached("proxy_probe_url", "https://api.openai.com/v1/models")
	if s, ok := probeURL.(string); ok {
		probeURL = s
	}
	targetURL, _ := probeURL.(string)
	if targetURL == "" {
		targetURL = "https://api.openai.com/v1/models"
	}

	route := services.NewRoute(&existing.URL, existing.LocalAddress)
	okRoute, latencyMs, statusCode, errStr := services.ProbeRoute(route, targetURL, nil, 15*time.Second)

	now := time.Now()
	existing.LastCheckedAt = &now
	existing.LatencyMs = &latencyMs

	if okRoute {
		if existing.Status == "disabled" {
			existing.Status = "active"
			existing.FailedCount = 0
			existing.DisabledReason = nil
		}
	} else {
		existing.FailedCount++
		maxFailures := services.GetInt("max_proxy_failures", 3)
		if existing.FailedCount >= maxFailures {
			existing.Status = "disabled"
			existing.DisabledReason = &errStr
		}
	}

	if err := db.Save(&existing).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update proxy status."})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"ok":          okRoute,
		"latency_ms":  latencyMs,
		"status_code": statusCode,
		"error":       errStr,
		"proxy":       proxyToOut(existing),
	})
}

func proxyToOut(p database.Proxy) models.ProxyOut {
	return models.ProxyOut{
		ID:             p.ID,
		URL:            p.URL,
		LocalAddress:   p.LocalAddress,
		Enabled:        p.Enabled,
		Status:         p.Status,
		FailedCount:    p.FailedCount,
		Weight:         p.Weight,
		LatencyMs:      p.LatencyMs,
		Note:           p.Note,
		DisabledReason: p.DisabledReason,
		LastUsedAt:     p.LastUsedAt,
		LastCheckedAt:  p.LastCheckedAt,
		CreatedAt:      p.CreatedAt,
	}
}

func RegisterProxyRoutes(router *gin.RouterGroup) {
	proxyGroup := router.Group("/proxies")
	{
		proxyGroup.GET("", listProxies)
		proxyGroup.POST("", batchAddProxies)
		proxyGroup.PATCH("/:id", updateProxy)
		proxyGroup.DELETE("/:id", deleteProxy)
		proxyGroup.POST("/:id/probe", probeProxy)
	}
}
